docs/modules/specifications/pages/decentral-nodes.adoc

```diff
@@ -0,0 +1,59 @@
+
+Each paragraph also aims to be as much as possible a standalone code change or addition, so be sure to use at least as many commits, and preferably at least a separate branch for each section.
+
+
+== HandlersJS extensions
+
+
+=== Storage
+
+We often use the same types of storage classes and abstractions, so we should centralize these in out **HandlersJS** repository. Create a directory called `storage` in the `handlersjs-core` library. Take a look at the `KeyValueStore` and its `(In)MemoryStore` implementation (leave out the 'In') in the `dgt-id-proxy` package of the **ID Broker** repository. Copy those classes to the new directory in **HandlersJS**, and export them as part of the public API.
+
+Code a `TypedKeyValueStore<M>` interface that extends from `KeyValueStore<keyof M, M[keyof M]>`. It should contain the same methods, but make them generic with a parameter `<T extends keyof M>`, and replace all key types `K` by `T` and value types `V` by `M[T]`. 
+
+Also code a `TimedKeyValueStore<K,V>` interface that extends from `KeyValueStore<K,V>`, and provides a function `latestUpdate(key: K)` returning a timestamp of when the value of that key has been updated, as well as a function `hasUpdate(key: K, time: number)` that returns a boolean depending on whether the time is older or newer than when the value of that key has been updated.
+
+Combine both interfaces in a `TimedTypedKeyValueStore<M>` interface.
+
+Make the `MemoryStore` implementation extends from the combined interface, implementing the new methods of the `TimedKeyValueStore`, and make it take an `initialData` parameter in it's construction, so we can add data via the ComponentsJS configuration (n.b. the initial data simply take the current time as latest update).
+
+
+=== Scheduler daemon
+
+We'll need to check peer nodes of a SemCom Node server repetitively, but we want to abstract the repetition pattern from the task that needs to be repeated. Start by moving the `Daemon` class from `handlersjs-http` to a `models` directory in the `handlersjs-core` library, and make the necessary changes to imports and exports. 
+
+Then add a `daemons` directory, an implement a `Scheduler` that extends `Daemon`. This class can take an `interval` (in milliseconds) and a any function as `task`. When started, this daemon will set a timeout to wait for the given interval and then execute the task and set a new timeout. It should track the running timeout, so that when stopped it can cancel it.
+
+
+=== Synchronization service
+
+Code a `SyncService` class, which takes a `TypedKeyValueStore`, the key for a `storage` set therein, another key for a set of `peers` URLs and an optional a string containing an `endpoint` suffix. This service tracks the time of the last synchronization in a variable, and should contain an asynchronous `sync` function which runs through all URLs in the list of peers with the following steps:
```

Comment:
```suggestion
Code a `SyncService` class, which takes a `TypedKeyValueStore`, the key for a `storage` set therein, another key for a set of `peers` URLs and an optional string containing an `endpoint` suffix. This service tracks the time of the last synchronization in a variable, and should contain an asynchronous `sync` function which runs through all URLs in the list of peers with the following steps:
```

Reply:
> Code a `SyncService` class, which takes a `TypedKeyValueStore`, the key for a `storage` set therein, another key for a set of `peers` URLs and an optional string containing an `endpoint` suffix

I'm also not sure I understand what is meant here, perhaps a method/constructor signature is easier to understand?

Reply:
I read this as: a store with 3 entries: (JSON representation of the Store)
```
{
  "storage": Set(as in the type)
  "peers": Set(as in the type),
  "endpoint": "some string",
}
```

Is that correct ?

Reply:
```suggestion
Code a `SyncService` class, which takes a `TypedKeyValueStore`, two keys (strings) `storage` and `peers`, and an optional a string containing an `endpoint` suffix. This service tracks the time of the last synchronization in a variable, and should contain an asynchronous `sync` function which runs through all URLs in the list of peers with the following steps:
```